dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EventPersister.java

```diff
@@ -172,63 +169,118 @@ private void handleDataValues(
+      TrackerPreheat preheat) {
+    EventDataValue eventDataValue = new EventDataValue();
+    eventDataValue.setDataElement(dataElement.getUid());
+    eventDataValue.setCreated(new Date());
+    eventDataValue.setCreatedByUserInfo(UserInfoSnapshot.from(user));
+    eventDataValue.setStoredBy(user.getUsername());
 
-          eventDataValue.setDataElement(dataElement.getUid());
-          eventDataValue.setStoredBy(dv.getStoredBy());
+    eventDataValue.setLastUpdated(new Date());
+    eventDataValue.setLastUpdatedByUserInfo(UserInfoSnapshot.from(user));
 
-          if (StringUtils.isEmpty(dv.getValue())) {
-            if (dataElement.isFileType()) {
-              unassignFileResource(
-                  entityManager, preheat, event.getUid(), eventDataValue.getValue());
-            }
+    eventDataValue.setValue(dv.getValue());
+    eventDataValue.setProvidedElsewhere(dv.isProvidedElsewhere());
 
-            event.getEventDataValues().remove(eventDataValue);
-          } else {
-            eventDataValue.setValue(dv.getValue());
+    if (dataElement.isFileType()) {
+      assignFileResource(entityManager, preheat, event.getUid(), eventDataValue.getValue());
+    }
 
-            if (dataElement.isFileType()) {
-              assignFileResource(entityManager, preheat, event.getUid(), eventDataValue.getValue());
-            }
+    event.getEventDataValues().add(eventDataValue);
+  }
 
-            event.getEventDataValues().remove(eventDataValue);
-            event.getEventDataValues().add(eventDataValue);
-          }
+  private void updateDataValue(
+      EventDataValue eventDataValue,
+      DataValue dv,
+      Event event,
+      DataElement dataElement,
+      UserDetails user,
+      EntityManager entityManager,
+      TrackerPreheat preheat) {
+    eventDataValue.setLastUpdated(new Date());
+    eventDataValue.setLastUpdatedByUserInfo(UserInfoSnapshot.from(user));
 
-          logTrackedEntityDataValueHistory(
-              user.getUsername(), dataElement, event, new Date(), valuesHolder);
-        });
+    if (dataElement.isFileType()) {
+      unassignFileResource(entityManager, preheat, event.getUid(), eventDataValue.getValue());
+      assignFileResource(entityManager, preheat, event.getUid(), dv.getValue());
+    }
+
+    eventDataValue.setProvidedElsewhere(dv.isProvidedElsewhere());
+    eventDataValue.setValue(dv.getValue());
   }
 
-  private Date getFromOrNewDate(DataValue dv, Function<DataValue, Instant> dateGetter) {
-    return Optional.of(dv).map(dateGetter).map(DateUtils::fromInstant).orElseGet(Date::new);
+  private void deleteDataValue(
+      EventDataValue eventDataValue,
+      Event event,
+      DataElement dataElement,
+      EntityManager entityManager,
+      TrackerPreheat preheat) {
+    if (dataElement.isFileType()) {
+      unassignFileResource(entityManager, preheat, event.getUid(), eventDataValue.getValue());
+    }
+
+    event.getEventDataValues().remove(eventDataValue);
```

Comment:
Can we rely on the equals implementation? I was bitten a few times by our equals implementations. Just wanted to double-check 😅

Reply:
In this case it is just checking the `dataElement` https://github.com/dhis2/dhis2-core/blob/accad7214119afaf3de7d702821216a2ba8c82de/dhis-2/dhis-api/src/main/java/org/hisp/dhis/eventdatavalue/EventDataValue.java#L120 so it works for us.